pkg/registry/rbac/clusterrole/policybased/storage_test.go

```diff
@@ -0,0 +1,253 @@
+limitations under the License.
+*/
+
+package policybased
+
+import (
+	"context"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/kubernetes/pkg/apis/rbac"
+	_ "k8s.io/kubernetes/pkg/apis/rbac/install"
+	"k8s.io/kubernetes/pkg/registry/rbac/validation"
+)
+
+func TestEscalation(t *testing.T) {
+	createContext := request.WithRequestInfo(request.WithNamespace(context.TODO(), ""), &request.RequestInfo{
+		IsResourceRequest: true,
+		Verb:              "create",
+		APIGroup:          "rbac.authorization.k8s.io",
+		APIVersion:        "v1",
+		Namespace:         "",
+		Resource:          "clusterroles",
+		Name:              "",
+	})
+	updateContext := request.WithRequestInfo(request.WithNamespace(context.TODO(), ""), &request.RequestInfo{
+		IsResourceRequest: true,
+		Verb:              "update",
+		APIGroup:          "rbac.authorization.k8s.io",
+		APIVersion:        "v1",
+		Namespace:         "",
+		Resource:          "clusterroles",
+		Name:              "myrole",
+	})
+
+	superuser := &user.DefaultInfo{Name: "superuser", Groups: []string{"system:masters"}}
+	superuserCreateCtx := request.WithUser(createContext, superuser)
+	superuserUpdateCtx := request.WithUser(updateContext, superuser)
+
+	bob := &user.DefaultInfo{Name: "bob"}
+	bobCreateCtx := request.WithUser(createContext, bob)
+	bobUpdateCtx := request.WithUser(updateContext, bob)
+
+	alice := &user.DefaultInfo{Name: "alice"}
+	aliceCreateCtx := request.WithUser(createContext, alice)
+	aliceUpdateCtx := request.WithUser(updateContext, alice)
+
+	var (
+		authzCalled   = 0
+		authzDecision = authorizer.DecisionNoOpinion
+		authzReason   string
+		authzErr      error
+	)
+	fakeStorage := &fakeStorage{}
+	fakeAuthorizer := authorizer.AuthorizerFunc(func(attr authorizer.Attributes) (authorizer.Decision, string, error) {
+		authzCalled++
+		return authzDecision, authzReason, authzErr
+	})
+	fakeRuleResolver, _ := validation.NewTestRuleResolver(
+		nil,
+		nil,
+		[]*rbac.ClusterRole{{ObjectMeta: metav1.ObjectMeta{Name: "alice-role"}, Rules: []rbac.PolicyRule{{APIGroups: []string{"*"}, Resources: []string{"*"}, Verbs: []string{"*"}}}}},
+		[]*rbac.ClusterRoleBinding{{RoleRef: rbac.RoleRef{Name: "alice-role", APIGroup: "rbac.authorization.k8s.io", Kind: "ClusterRole"}, Subjects: []rbac.Subject{{Name: "alice", Kind: "User", APIGroup: "rbac.authorization.k8s.io"}}}},
+	)
+
+	role := &rbac.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{Name: "myrole", Namespace: ""},
+		Rules:      []rbac.PolicyRule{{APIGroups: []string{""}, Verbs: []string{"get"}, Resources: []string{"pods"}}},
+	}
+
+	s := NewStorage(fakeStorage, fakeAuthorizer, fakeRuleResolver)
+
+	// Superuser
```

Comment:
Might help the upkeep of these tests to add a little meat to the comments (like if we decide to switch the order of the escalation checks or something).

```
// Superuser: Can create/update role w/o escalation, so authorizer is not called, and create/update call should succeed.

// Unauthorized: Bob must escalate to create/update the role. When he doesn't have "escalate" permission, the calls should fail.

// Authorized: Bob must escalate to create/update the role. When he does have "escalate permission, the calls should succeed.

// Implicitly authorized: Alice ...
```
Actually, I wasn't sure why the Alice thing worked the way it did. I'll take another look in a little bit.

Reply:
made it a table test, added more descriptions

Reply:
Awesome. That's even better.